[WFLY-10493]: Support JMS Java EE resources definition for remote Artemis-based broker.

ehsavoie opened the pull request:
* Creating queue / topic on deployments
    * Deleting queue / topic in undeployments
    * Looking for default ee resource adapter if none is specified.
    * Making @JMSConnectionFactoryDefinition work without internal broker.
    * Adding support for description of runtime external destinations.
    * Adding support for management address definition in the @JMSDestinationDefinition

 JIRA: https://issues.jboss.org/browse/WFLY-10493 
https://github.com/wildfly/wildfly-proposals/pull/78

ehsavoie committed:
[WFLY-10493]: Support JMS Java EE resources definition for remote Artemis-based broker.

* Creating queue / topic on deployments
* Deleting queue / topic in undeployments
* Looking for default ee resource adapter if none is specified.
* Making @JMSConnectionFactoryDefinition work without internal broker.
* Adding support for description of runtime external destinations.
* Adding support for management address definition in the @JMSDestinationDefinition.
* Fixing WFWIP-147: @JMSConnectionFactoryDefinition with property resourceAdapater to pooled-connection-factory defined in messaging-active subsystem does not work.
* Adding support for authentication of management queue.

JIRA: https://issues.jboss.org/browse/WFLY-10493

bstansberry commented:
@ehsavoie This will need a community docs commit as well.

jmesnil commented on messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/jms/ExternalJMSQueueService.java:
```diff
@@ -36,19 +53,94 @@
+            final NamingContext storeBaseContext = new NamingContext(namingStore, null);
+            try {
+                QueueConnectionFactory cf = (QueueConnectionFactory) storeBaseContext.lookup(pcfInjector.getValue().getBindInfo().getAbsoluteJndiName());
+                try (QueueConnection connection = cf.createQueueConnection()) {
```
what if the remote Artemis has enabled authentication on the management address?

ehsavoie replied:
Good point

ehsavoie commented on messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/deployment/JMSConnectionFactoryDefinitionInjectionSource.java:
```diff
@@ -330,12 +375,44 @@ static boolean targetsPooledConnectionFactory(String server, String resourceAdap
+        return serviceRegistry.getServiceNames().contains(pcfName);
+    }
+
+    static String getDefaulResourceAdapter(DeploymentUnit deploymentUnit) {
```
yes :)

ehsavoie commented on messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/ServerDefinition.java:
```diff
@@ -847,5 +845,27 @@ public void registerChildren(ManagementResourceRegistration resourceRegistration
+        public static final String validValues;
 
+        static {
+            StringJoiner sj = new StringJoiner(",");
```
because it was in Artemis code ;)

ehsavoie commented on docs/src/main/asciidoc/_admin-guide/subsystem-configuration/Messaging_Connect_a_pooled-connection-factory_to_a_Remote_Artemis_Server.adoc:
```diff
@@ -118,3 +144,24 @@ http://activemq.apache.org/artemis/docs/1.1.0/using-jms.html#jndi-configuration[
+[[configuration of a temporary remote destination]]
+==== Configuration temporary remote destination
+
+The annotation `@JMSDestinationDefinition` can be used to create temporary destination on a remote Artemis Server. This will work in the same way as for a local server.+
```
I meant that the destination existence is bound to the deployment lifecycle

ehsavoie committed:
[WFLY-10493]: Support JMS Java EE resources definition for remote Artemis-based broker.

* Simplifying code and reducing coupling with Artemis.
* Fixing WFWIP-149: creating external queues in a remote broker cluster.
Jira: https://issues.jboss.org/browse/WFLY-10493
      https://issues.jboss.org/browse/WFWIP-149
